lib/Dialect/FIRRTL/FIRRTLFolds.cpp

```diff
@@ -67,6 +67,14 @@ static bool hasKnownWidthIntTypesAndNonZeroResult(Operation *op) {
   return true;
 }
 
+/// Return true if this value is 1 bit UInt.
+static bool valueIsOneBitUInt(Value op) {
+  auto t = op.getType().cast<UIntType>();
+  if (!t || !t.hasWidth() || t.getWidth() != 1)
+    return false;
+  return true;
+}
```

Comment:
I think `cast` is guaranteed to not return null, otherwise it will assert.  Looks like the ops also support signed ints.

Reply:
Yeah, this should always be an `IntType` so either `UInt` or `SInt` can show up here. And the optimization works for `SInt`, too.

Reply:
I can't generate a SInt to UInt cast durring fold, which is what would be needed for SInt reductions.
These ops are constrained to int types, which is why the cast is safe.  It's result is checked as it may be an SInt which isn't supported in the fold.

Reply:
It would be nice to catch the SInt case in a canonicalization.

Reply:
I see what you mean about the casts, good call on that.  I think this should have been a `dyn_cast` though, its crashing if you use a `sint`:
```mlir
// ./bin/firtool test.mlir
// Assertion failed: (isa<U>()), function cast, file ../llvm/llvm/../mlir/include/mlir/IR/Types.h, line 243.
// firtool`circt::firrtl::AndRPrimOp::fold(this=0x00007ffeefbfb0c0, operands=ArrayRef<mlir::Attribute> @ 0x00007ffeefbfaf60) at FIRRTLFolds.cpp:727:7 [opt]
firrtl.circuit "Reduce" {
firrtl.module @Reduce(in %a: !firrtl.sint<1>, out %b: !firrtl.uint<1>) {
  %0 = firrtl.andr %a : (!firrtl.sint<1>) -> !firrtl.uint<1>
  firrtl.connect %b, %0 : !firrtl.uint<1>, !firrtl.uint<1>
}
}
```

Reply:
e7002ad3

Reply:
:+1: Thanks, Chris!